velox/exec/GroupingSet.cpp

```diff
@@ -275,13 +275,15 @@ void GroupingSet::initializeGlobalAggregation() {
 
   // Row layout is:
   //  - null flags - one bit per aggregate,
-  // - uint32_t row size,
+  //  - uint32_t row size,
   //  - fixed-width accumulators - one per aggregate
   //
   // Here we always make space for a row size since we only have one
   // row and no RowContainer.
   int32_t rowSizeOffset = bits::nbytes(aggregates_.size());
   int32_t offset = rowSizeOffset + sizeof(int32_t);
+  // Accumulators must be aligned. Use 16 bytes to accommodate int128_t types.
```

Comment:
Why need this alignment? The offset advance at l296 doesn't guarantee this alignment for non-first accumulator?

Reply:
Line 291 already sets the offset of the first accumulator. Line 296 simply increments the current offset by the fixed width size of the accumulator.  
If you have say 5 values, offset becomes `5 (1 + 4)` for the first accumulator, and the offset gets incremented from `5` for the remaining accumulators.

Reply:
I might miss some context here. My question is that this can only guarantee the offset passed to setOffsets at l291 to be aligned for the first aggregator? It can still be unaligned for the other aggregators? Why we need to guarantee the offset to be aligned for the first aggregator but not the others? Thanks!

Reply:
I fixed this. I was optimistic with my assumption that `accumulatorFixedWidthSize()` will return an aligned size.

Reply:
Thanks and do we need to update the offset setting in RowContainer for non-global aggregation case as well? I am also curious that the accumulator couldn't handle the mis-aligned access using reinterpret cast? Otherwise, the 16 bytes alignment by default might waste some memory in case of non-global aggregation. Thanks!

Reply:
The current flow is 
https://github.com/facebookincubator/velox/blob/ff331272ad5f67ad5585085573c3401528282de1/velox/functions/prestosql/aggregates/SumAggregate.h#L54-L60
where
https://github.com/facebookincubator/velox/blob/ff331272ad5f67ad5585085573c3401528282de1/velox/exec/Aggregate.h#L261-L264

All aggregations have a similar flow. Without this alignment, I saw a segfault for aggregations involving `int128_t` types even with the above `reinterpret_cast`. Unaligned accesses in general are not portable across architectures. They could also lead to performance issues in terms of additional load instructions.

I did not realize RowContainer layout also has accumulators. We should fix this there as well. Do you want me to add this fix there?

Reply:
Yes, please. Also I have chat this with @mbasmanova offline, is it possible to let specific accumulator to reserve additional memory if it needs aligned memory access and accumulator actually reports the actual bytes to use through  accumulatorFixedWidthSize? I thought using reinterpret_cast<T*>(group + offset_) should work? Have you checked the memory address that caused the segment fault? Then we might need to provide a template boolean flag say ALIGNED and return a pointer which is aligned to type T for those type of accumulator.

Reply:
https://github.com/facebookincubator/velox/pull/2163 introduces an int128_t type (LongDecimal) accumulator and you can reproduce this crash there without this alignment change. Yes, I verified the memory address and it was not aligned to 16 bytes. Introducing the alignment fixed the crash.
Making the alignment specific to the accumulator can be done. However, adding it to the accumulatorFixedWidthSize will end up wasting more space. As @Yuhta suggested above, we can add another virtual function for alignment size.
Since the scope is now large, I will remove this change from this PR and I will add a new PR to handle the alignment.